text/0033-input-versions.md

```diff
@@ -0,0 +1,209 @@
+
+This RFC proposes to define annotations that signal the associated version of a particular input format, so that future versions of Cedar (or tools/services that use Cedar) can handle the situation gracefully.
+
+This RFC makes no specification about what "gracefully" means, but minimally it can be used to indicate that parse errors might be due to version format incompatibilities. For example, by annotating a policy or JSON input file with version 2.4, the parser for Cedar version 3.0 can flag an input with a 2.4-only feature as no longer supported, rather than just failing to parse. Or, a service that uses Cedar can run multiple versions of the Cedar tools, selecting the version that matches the input version they are given.
+
+## Detailed design
+
+This RFC proposes a version annotation on Cedar's main input types: policies, entities (JSON), schema (JSON and custom). 
+
+The design has two main goals:
+
+- **Optional**: No version annotation is required. If none is given, the Cedar parser will assume the most recent version.
+
+- **Uniform**: The way of specifying the version should be similar across all input formats, but also adhere to their particular style.
+
+### Version string
+
+Each input file will be associated with a version string. This is the version the input was written for, e.g., `"2.4.1"` or `"3.0"`. It could be that newer Cedar versions are compatible with older formats, but it is up to newer parsers to know that. For example, a new version of Cedar will add the [`is` operator](0005-is-operator.md) to policies, but an earlier version without that operator may still parse properly. The newer parser can account for this when parsing.
+
+If an input file is annotated with an ill-formed version string, the parser should treat it as a version it does not know about, e.g., as if for a future version of Cedar. It can issue a warning and proceed. If an input file is version-annotated more than once, then the parser should signal an error.
+
+### Policies and Custom Schema
+
+For Cedar policies and custom schema syntax, we propose adding a new top-level declaration:
+```
+@cedar_version(XXX);
+```
+where `XXX` is a string with the version in it.
+
+This declaration applies to all policies or schema elements that are parsed in the same file. It doesn't matter where the declaration is, in the file. If the declaration appears more than once, it is an error.
+
+Here is the amended grammar for a policies file:
+```
+File    := {Policy | Version}
+Version := '@cedar_version' '(' STRING ')' ';'
+Policy  := {Annotation} Effect '(' Scope ')' {Conditions} ';'
+...
+```
+Here is the amended grammar for a custom schema:
+```
+Schema    := {Namespace | Version}
+Version   := '@cedar_version' '(' STRING ')' ';'
+Namespace := ('namespace' Path '{' {Decl} '}') | {Decl}
+...
+```
+In both cases there is the semantic constraint that `Version` appears at most once.
+
+### Entities JSON
```

Comment:
Not mentioned in this RFC: if we have explicit version markers for the entities JSON format, would we also have/need explicit version markers for entities constructed programmatically (without JSON)?  I would say this RFC should take a stance on this (yes/no), and if yes, then at least sketch how the APIs would change.

This might be an additional argument in support of some kind of global configuration rather than per-file or more granular version markers (as discussed in other threads, e.g., [this one](https://github.com/cedar-policy/rfcs/pull/33#discussion_r1381889216), [this one](https://github.com/cedar-policy/rfcs/pull/33#discussion_r1381930212), or [this one](https://github.com/cedar-policy/rfcs/pull/33#issuecomment-1773893005)).